cupy/core/_routines_math.pyx

```diff
@@ -75,6 +75,12 @@ cdef ndarray _ndarray_imag_setter(ndarray self, value):
 
 
 cdef ndarray _ndarray_prod(ndarray self, axis, dtype, out, keepdims):
+    if cupy.cuda.cub_enabled:
+        # result will be None if the reduction is not compatible with CUB
+        result = cub.cub_reduction(self, cub.CUPY_CUB_PROD, axis, dtype, out,
+                                   keepdims)
```

Comment:
Is the dtype promoted in cub?

Reply:
No, it uses the same rule as in CUB-backed `sum()`, see the function `cupy.cuda.cub. _cub_reduce_dtype_compatible()`. If the dtype promotion is needed, currently we fall back to the old kernels (without CUB).

I am not sure if we want to handle type promotion within the CUB module. Since we don't write our own kernels in this case, to promote dtype an extra copy needs to be done before passing it to the CUB kernel. I think this was probably why we took the current approach.

Reply:
in numpy and cupy, the current prod kernel returns an `int64` array if the input is `int32`.

So with cub we can't replicate this behavior, right?

Reply:
No, we can’t do it with the CUB kernel, so in this case CuPy’s prod ufunc will be invoked to do the type conversion (`i->l`):
https://github.com/cupy/cupy/blob/de3d32536c39eefea1805f4a7a76a14356eb3bff/cupy/core/_routines_math.pyx#L576